modules/failure-handler/src/main/java/org/apache/ignite/internal/failure/FailureManager.java

```diff
@@ -225,4 +271,36 @@ public synchronized void setInterceptor(@Nullable FailureHandler interceptor) {
     FailureHandler handler() {
         return handler;
     }
+
+    /**
+     * Defines whether thread dump should be throttled for given failure type or not.
+     *
+     * @param type Failure type.
+     * @return {@code true} if thread dump generation should be throttled for given failure type.
+     */
+    private boolean throttleThreadDump(FailureType type) {
+        Map<FailureType, Long> dumpPerFailureTypeTs = threadDumpPerFailureTypeTs;
+        long dumpThrottlingTimeout = dumpThreadsThrottlingTimeout;
+
+        if (dumpThrottlingTimeout == 0 || dumpPerFailureTypeTs == null) {
+            return false;
+        }
+
+        long curr = System.currentTimeMillis();
+
+        Long last = dumpPerFailureTypeTs.get(type);
+
+        assert last != null : "Unknown failure type " + type;
+
+        boolean throttle = curr - last < dumpThrottlingTimeout;
+
+        if (!throttle) {
+            dumpPerFailureTypeTs.put(type, curr);
```

Comment:
`EnumMap` is not thread safe. How is this going to work?

Reply:
The `throttleThreadDump` is called by `process` method which is `synchronized`. So, it looks OK.

Reply:
Does it make sense to say in its javadoc that it should be called under synchronization?

Reply:
Updated.